TELCODOCS-2306: KMM 2.4 Release Notes (#95248)
openshift/openshift-docs: hardware_enablement/kmm-release-notes.adoc

* In this release, KMM uses version 1.23 of the Golang programming language to ensure test continuity for partners.

// TELCODOCS-2197
* You can now schedule KMM pods by defining taints and tolerations. For more information, see xref:../hardware_enablement/kmm-kernel-module-management.adoc#kmm-using-tolerations-for-kernel-module-scheduling_kernel-module-management-operator[Using tolerations for kernel module scheduling].

[id="kmm-2-4-RN"]
== Release notes for Kernel Module Management Operator 2.4
=== New features and enhancements
// TELCODOCS-2311

> * In this release, you now have the option to configure the KMM module to not load an out-of-tree kernel driver and use the in-tree driver instead, and run only the device plugin. For more information see xref:../hardware_enablement/kmm-kernel-module-management.adoc#kmm-using-intree-modules_kernel-module-management-operator[Using in-tree modules].

Author: Made these changes.

// TELCODOCS-2304

> * In this release, KMM configurations are now persistent following cluster and KMM Operator upgrades and redeployments of KMM. 

Reviewer: ```suggestion
* In this release, KMM configurations are now persistent after cluster and KMM Operator upgrades and redeployments of KMM. 
```

Author: changed

+

> In earlier releases, a cluster or KMM upgrade, or any other intentional or unintentional action, such as upgrading a non-default configuration like the firmware path that redeploys KMM, could create the need to reconfigure KMM. In this release, KMM configurations now remain persistent regardless of any of such actions.

Author: Good point. changed.

+
For more information, see xref:../hardware_enablement/kmm-kernel-module-management.adoc#kmm-configuring-kmmo_kernel-module-management-operator[Configuring the Kernel Module Management Operator].

// MGMT-19735
* Improvements have been added to KMM so that GPU Operator vendors will not need to replicate KMM functionality in their code, but instead use KMM as is. This change greatly improves Operators' code size, tests, and reliability.

// MGMT-18966
* In this release, KMM no longer uses HTTP(S) direct requests to check if a kmod image exists. Instead, CRI-O is used internally to check for the images. This mitigates the need to access container image registries directly from HTTP(S) requests and manually handle tasks such as reading `/etc/containers/registries.conf` for mirroring configuration, accessing the image cluster resource for TLS configuration, mounting the CAs from the node, and maintaining our own cache in Hub & Spoke. 

// MGMT-19383

> The KMM and KMM-hub Operators have been assigned the "Meets Best Practices" label in the 

Author: changed

https://catalog.redhat.com/search?searchType=software[Red Hat Catalog].

// MGMT20613

> * You can now install KMM on worker nodes, if needed. Previously, it was not possible to deploy workloads on the control-plane nodes. Because the worker nodes do not have the `node-role.kubernetes.io/control-plane` or `node-role.kubernetes.io/master` labels, the Kernel Module Management Operator might need further configurations. An internal code change has resolved this issue.  

Author: changed

* In this release, KMM uses version 1.23 of the Golang programming language to ensure test continuity for partners.

// TELCODOCS-2197
* You can now schedule KMM pods by defining taints and tolerations. For more information, see xref:../hardware_enablement/kmm-kernel-module-management.adoc#kmm-using-tolerations-for-kernel-module-scheduling_kernel-module-management-operator[Using tolerations for kernel module scheduling].

[id="kmm-2-4-RN"]
== Release notes for Kernel Module Management Operator 2.4
=== New features and enhancements
// TELCODOCS-2311

> * In this release, you now have the option to configure KMM Module to not load an out-of-tree kernel driver and use the in-tree driver instead and run only the device plugin. For more information see xref:../hardware_enablement/kmm-kernel-module-management.adoc#kmm-using-intree-modules_kernel-module-management-operator[Using in-tree modules].

Author: changed.

// TELCODOCS-2304
* In this release, KMM configurations are now persistent following cluster and KMM Operator upgrades and redeployments of KMM. 
+

> In earlier releases, a cluster or KMM upgrade, or any other intentional or unintentional action, such as upgrading a non-default configuration like the firmware path that redeploys KMM, could create the need to reconfigure KMM. In this release, KMM configurations now remain persistent regardless of any of these actions.

Author: "intentional or unintentional" is important. Leaving it in. Made other change.

+

> For more information see xref:../hardware_enablement/kmm-kernel-module-management.adoc#kmm-configuring-kmmo_kernel-module-management-operator[Configuring the Kernel Module Management Operator].

Author: added.

// MGMT-19735

> * Improvements have been added to KMM so that GPU Operator vendors will not need to replicate KMM functionality in their code, but instead use KMM as is. This will greatly improve Operators' code size, tests, and reliability.

Author: changed.

// MGMT-18966

> * In this release, KMM no longer uses HTTP[s] direct requests to check if a kmod image exists. Instead, CRI-O is used internally to check for the images. This mitigates the need to access container image registries directly from HTTP[s] requests and manually handle tasks such as reading `/etc/containers/registries.conf` for mirroring configuration, accessing the image cluster resource for TLS configuration, mounting the CAs from the node, and maintaining our own cache in Hub & Spoke. 

Author: Made these changes.

// MGMT-19383

> The KMM and KMM-hub Operators have been assigned the "Meets Best Practices" label in the 

Reviewer 2: ```suggestion
* The KMM and KMM-hub Operators have been assigned the "Meets Best Practices" label in the 
```

https://catalog.redhat.com/search?searchType=software[Red Hat Catalog].

// MGMT20613

> * You can now install KMM on worker nodes if there is no better choice. Previously, it was not allowable to deploy workloads on the control-plane nodes. Because the worker nodes do not have the `node-role.kubernetes.io/control-plane` or `node-role.kubernetes.io/master` labels the Kernel Module Management Operator will not run out of the box. An internal code change has resolved this issue.  

Author: s/if there is no better choice/, if needed
s/allowable/possible

// MGMT-20249
* In this release, the heartbeat filter for the NMC reconciler has been updated to filter the following events on nodes:

> ** node spec

Author: changed

> ** metadata.labels

Author: changed

> ** status.nodeInfo

Author: changed

> ** status.conditions[] (NodeReady only) and still filtering heartbeats  

Author: changed

== Notable technical changes
// TELCODOCS-2343
* In this release, the preflight validation resource in the cluster has been modified. You can use preflight validation to verify kernel modules to be installed on the nodes after cluster upgrades and possible kernel upgrades. Preflight validation also reports on the status and progress of each module in the cluster that it attempts or has attempted to validate. For more information, see xref:../updating/preparing_for_updates/kmm-preflight-validation.adoc#kmm-validation-kickoff_kmm-preflight-validation[Preflight validation for Kernel Module Management (KMM) Modules].

// TELCODOCS-2344
* A requirement when creating a kmod image is that both the `.ko` kernel module files and the `cp` binary must be included, which is required for copying files during the image loading process.
For more information, see xref:../hardware_enablement/kmm-kernel-module-management.adoc#kmm-creating-kmod-image_kernel-module-management-operator[Creating a kmod image].

//MGMT-19919

> * The `capabilities` field that refers to the Operator maturity level has been changed from `Basic Install` to `Seamless upgrades`. `Basic Install` indicates that the Operator doesn't have an upgrade option. This is not the case for KMM, where seamless upgrades are supported.

Author: changed

=== Bug fixes
// MGMT-19548
* Webhook deployment has been renamed from `webhook-server` to `webhook`.

** *Cause*: Generating files with `controller-gen` generated a service called `webhook-service` that is not configurable. And, when deploying KMM with OLM, OLM deploys a service for the webhook called `-service`.

> ** *Consequence*: Two services were generated for the same deployment. One generated by `controller-gen` and added to the bundle manifests and the other one created by OLM.

Author: changed

> ** *Consequence*: Two services were generated for the same deployment. One generated by `controller-gen` and added to the bundle manifests and the other one created "on the fly" by OLM.

Author: deleted "on the fly".

> ** *Fix*: Make OLM find an already existing service called `webhook-service` in the cluster because the deployment is called webhook, 

Author: changed

** *Result*: A second service is no longer created.

// MGMT-20892 
* Using `imageRepoSecret` in conjunction with DTK as imagestream results in `authorization required` error.

> ** *Cause*: On the Kernel Module Management (KMM) Operator, when you set `imageRepoSecret` in the KMM module, and the build's resulting container image is defined to be stored in the cluster's internal registry, the build will fail to push the final image and generate an `authorization required` error. 

Author: changed

** *Consequence*: The KMM Operator does not work as expected. 

** *Fix*: When the `imageRepoSecret` is user-defined, it is used as both a pull and push secret by the build process. In order to support using the cluster's internal regsitry, the auth token for that registry needs to be added to the `imageRepoSecret`. The token can be obtained from the "build" service account of the KMM module's namespace. 

** *Result*: The KMM Operator works as expected. 

// MGMT-16797
* Creating or deleting the image / Creating an MCM module does not load the module on the spoke. 

> ** *Cause*: On a Hub/Spoke setup, when creating or deleting the image in registry, or when creating a `ManagedClusterModule` (MCM), the module on the spoke cluster is not loaded.

Author: changed

** *Consequence*: The module on the spoke is not created.

> ** *Fix*: Remove the cache package and image translation from the Hub/Spoke setup.

Author: changed

** *Result:* The module on the spoke is created for the second time the MCM object is created.

// MGMT-19859
* KMM cannot pull images from the private registry while doing in-cluster builds.

** *Cause*: The Kernel Module Management (KMM) Operator cannot pull images from private registry while doing in-cluster builds.

** *Consequence*: Images in private registries that are used in the build process can not be pulled.

** *Fix*: The `imageRepoSecret` configuration is now also used in the build process. The `imageRepoSecret` specified has to include ALL registries being used.

> ** *Result:* Private registries can now be used when doing in-cluster builds.

Author: changed

// MGMT-19897
* KMM worker pod is orphaned when deleting a module with a container image that can not be pulled. 

** *Cause*: A Kernel Module Management (KMM) Operator worker pod is orphaned when deleting a module with a container image that can not be pulled.

** *Consequence*: Failing worker pods are left on the cluster and at no point being collected for garbage.

** *Fix*: KMM will now garbage collect orphaned failing pods upon the modules deletion.

** *Result:* The module is successfully deleted, and all associated orphaned failing pods are also deleted.

// MGMT-19898
* The KMM Operator tries to create a MIC even when the node selector does not match.

** *Cause*: The Kernel Module Management (KMM) Operator tries to create a 'ModuleImagesConfig' (MIC) resource even when the node selector does not match with any actual nodes and fails.

** *Consequence*: The KMM Operator is getting and error when reconciling a module that does not target any node.

> ** *Fix*: The 'Images' field in the MIC resource is now optional.

Author: changed

** *Result:* The KMM Operator can successfully create the MIC resource even when there are no images in it.

// MGMT-20247
* KMM does not reload the kernel module in case the node reboot sequence is too quick.

** *Cause*: The Kernel Module Management (KMM) Operator does not reload the kernel module in case the node reboot sequence is too quick. The reboot is determined based on the timestamp of the status condition being later than the timestamp in the Node Machine Configuration (NMC) status.

> ** *Consequence*: When the reboot happens quickly, in less time than the grace period, the node state will not change. After the node reboots, KMM will not load the kernel module again.

Author: changed

** *Fix*: Instead of relying on the condition state, NMC can rely on the `Status.NodeInfo.BootID` field. This field is set by kubelet based on the /proc/sys/kernel/random/boot_id file of the server node, so it is updated after each reboot.

** *Result:* The more accurate timestamps enable the Kernel Module Management (KMM) Operator to reload the kernel module after the node reboot sequence.

// MGMT-20248

> * Filtering out node heartbeats events for the NMC controller.

Author: Added.

> ** *Cause*: The Node Machine Configuration (NMC) controller gets spammed with events from node heartbeats. The node heartbeats are there to let the Kubernetes API server know that the node is still connected and functional.

Author: changed

> ** *Consequence*: The spamming causes a constant reconciliation even when no Module, and therefore no NMC, are applied to the cluster.

Author: changed

> ** *Fix*: The NMC controller will now filter the node's heartbeat from its reconciliation loop.

Author: changed

** *Result:* The NMC controller only gets real events and filters out node heartbeats.

// MGMT-20286

> * NMC status contains toleration values, even though there are no tolerations in the NMC.spec or in the Module. 

Author: changed

** *Cause*: The Node Machine Configuration (NMC) status contains toleration values, even though there are no tolerations in the NMC.spec or in the Module.

** *Consequence*: Tolerations other than Kernel Module Management-specific tolerations can appear in the status.

** *Fix*: The NMC status now gets its toleration from a dedicated annotation rather than from the worker pod.

** *Result:* The NMC status only contains the Module's tolerations.

// MGMT-20725

> * The KMM Operator version 2.4 fails to start properly and cannot list the resource `\modulebuildsignconfigs\`. 

Author: changed

> ** *Cause*: On the Kernel Module Management (KMM) Operator, when the Operator is installed using Red Hat Konflux it does not start properly as the log files contain errors.

Author: changed

** *Consequence*: The KMM Operator does not work as expected.

> ** *Fix*: The Cluster Service Version (CSV) file was updated to list the resources `\modulebuildsignconfigs\` and `moduleimagesconfig`.

Author: changed

** *Result:* The KMM Operator works as expected.

// MGMT-20752

> * Konflux build does not include version and git commit in the operator logs.

Author: changed

** *Cause*: On the Kernel Module Management (KMM) Operator, when the Operator was built using Communications Platform as a Service (CPaas), the build included the Operator version and git commit in the log  files. However, with Red Hat Konflux these details are not included in the log files.

** *Consequence*: Important information is missing from the log files.

> ** *Fix*: Some adjustments were made in Konflux.

Author: changed

> ** *Result:* The KMM Operator build now includes the Operator version and git commit in the log files.

Reviewer 2: ```suggestion
** *Result:* The KMM Operator build now includes the Operator version and git commit ID in the log files.
```

Author: changed

> // MGMT-20754

Reviewer 2: Duplicate entry of teh same issue [here](https://github.com/openshift/openshift-docs/pull/95248/files#diff-a508e43343db975c5b8b7e3b4f30ef90441b9420f08df906454a6d85aa693514R144)

* The KMM Operator does not load the module after node with taint is rebooted.

** *Cause*: The Kernel Module Management (KMM) Operator does not reload the kernel module in case the node reboot sequence is too quick. The reboot is determined based on the timestamp of the status condition being later than the timestamp in the Node Machine Configuration (NMC) status.

> ** *Consequence*: When the reboot happens quickly, in less time than the grace period, the node state will not change. After the node reboots, KMM will not load the kernel module again.

Reviewer 2: ```suggestion
** *Consequence*: When the reboot happens quickly, in less time than the grace period, the node state does not change. After the node reboots, KMM does not load the kernel module again.
```

Author: changed.

** *Fix*: Instead of relying on the condition state, NMC can rely on the `Status.NodeInfo.BootID` field. This field is set by kubelet based on the /proc/sys/kernel/random/boot_id file of the server node, so it is updated after each reboot.

> ** *Result:* The more accurate timestamps enable the Kernel Module Management (KMM) Operator to reload the kernel module after the node reboot sequence.

Reviewer 2: Delete duplicate content

// MGMT-20775
* Redeploying a module that uses in-cluster builds fails with the `ImagePullBackOff` policy.

> ** *Cause*: On the Kernel Module Management (KMM) Operator, the `pullPolicy` for the puller pod and the worker pod is different.

Reviewer 2: ```suggestion
** *Cause*: On the Kernel Module Management (KMM) Operator, the image pull policy for the puller pod and the worker pod is different.
```

> ** *Consequence*: An image can be considered as existing while it fact it is not.

Reviewer 2: ```suggestion
** *Consequence*: An image can be considered as existing while, in fact, it is not.
```

Author: changed

> ** *Fix*: Make the pull policy of the pull pod the same as the pull policy defined in the KMM Module, as this is what is used by the worker pod.

Reviewer 2: ```suggestion
** *Fix*: Make the image pull policy of the pull pod the same as the pull policy defined in the KMM module since its the same policy that is used by the worker pod.
```

Author: changed

> ** *Result:* The MIC will represent the state of the image in the same way the worker pod will be able to access it.

Reviewer 2: ```suggestion
** *Result:* The MIC represents the state of the image in the same way the worker pod accesses it.
```

// MGMT-20785
* The MIC controller creates two pull-pods when it should create just one.

** *Cause*: On the Kernel Module Management (KMM) Operator, the `ModuleImagesConfig` (MIC) controller may create multiple pull-pods for the same image.

> ** *Consequence*: Resources are wasted and confusion is caused.

Reviewer 2: ```suggestion
** *Consequence*: Resources are not used appropriately or as intended.
```

** *Fix*: The `CreateOrPatch` MIC API receives a slice of `ImageSpecs`, as the input is created by going over the the target nodes and adding their images to the slice, so any duplicate `ImageSpecs`, are now filtered out.

** *Result:* The KMM Operator works as expected.

// MGMT-20827

> * The `job.dcDelay` example in the documentation should specify `0s` instead of 0.

Reviewer 2: ```suggestion
* The `job.dcDelay` example in the documentation should specify `0s` instead of `0`.
```

> ** *Cause*: The Kernel Module Management (KMM) Operator default `job.gcDelay` duration field is 0s but the documentation says it is `0`.

Reviewer 2: ```suggestion
** *Cause*: The Kernel Module Management (KMM) Operator default `job.gcDelay` duration field is `0s` but the documentation mentions the value as `0`.
```

> ** *Consequence*: This can lead a customer in the wrong direction setting a custom value of 60 instead of 60s or 1m that will resolve in an error due to the wrong input type.

Reviewer 2: ```suggestion
** *Consequence*: Entering a custom value of `60` instead of `60s` or `1m` might result in an error due to the wrong input type.
```

> ** *Fix*: The `job.gcDelay` in the documentation is updated to have a default value of 0s.

Reviewer 2: ```suggestion
** *Fix*: The `job.gcDelay` field in the documentation is updated to default value of `0s`.
```

Author: changed

** *Result:* Users are less likely to get confused.

// MGMT-20835

> * The KMM Operator Hub environment does not work because of missing MIC and MBSC CRDs.

Reviewer 2: Expand MBSC on first use , use abbreviation later
```suggestion
* The KMM Operator Hub environment does not work because of missing MIC and MBSC CRDs.
```

> ** *Cause*: The Kernel Module Management (KMM) Operator Hub environment only generates Custom Resource Definitions (CRD) files based on the 'api-hub/' directory. As a result, this does not contain some CRDs required for the KMM Operator Hub environment, such as, 'ModuleImagesConfig' (MIC) resource and Managed Kubernetes Service (MBSC).

Reviewer 2: ```suggestion
** *Cause*: The Kernel Module Management (KMM) Operator hub environment only generates Custom Resource Definitions (CRD) files based on the `api-hub/` directory. As a result, this does not contain some CRDs that are required for the KMM Operator Hub environment, such as, `ModuleImagesConfig` (MIC) resource and Managed Kubernetes Service (MBSC).
```

Author: changed

> ** *Consequence*: The KMM Operator Hub environment can not work because it tries to start controllers reconciling CRDs that do not exist in the cluster.

Reviewer 2: ```suggestion
** *Consequence*: The KMM Operator hub environment cannot work because it tries to start controllers reconciling CRDs that do not exist in the cluster.
```

> ** *Fix*: The fix generates all CRD files into the 'config/crd-hub/bases' directory, but only applies the resources to the cluster that it actually needs.

Reviewer 2: ```suggestion
** *Fix*: The fix generates all CRD files into the `config/crd-hub/bases` directory, but only applies the resources to the cluster that it actually needs.
```

Author: changed

> ** *Result:* The KMM Operator Hub environment works as expected.

Reviewer 2: ```suggestion
** *Result:* The KMM Operator hub environment works as expected.
```

// MGMT-20852

> * The KMM Operator Hub environment cannot build when finalisers can not be set on a resource.

Reviewer 2: Suggestion based on the assumtion that this refers to [OperatorHub](https://operatorhub.io/about) and not the hub-spoke of KMM
```suggestion
* The KMM OperatorHub environment cannot build when finalisers are not set on a resource.
```

> ** *Cause*: The Kernel Module Management (KMM) Operator displays an error with the 'ManagedClusterModule' controller failing to build. This is due to the missing 'ModuleImagesConfig' (MIC) resource finalizers and Role-based Action Control (RBAC) permissions for the KMM Operator Hub environment.

Reviewer 2: ```suggestion
** *Cause*: The Kernel Module Management (KMM) Operator displays an error with the `ManagedClusterModule` controller failing to build. This is due to the missing `ModuleImagesConfig` (MIC) resource finalizers and Role-based Action Control (RBAC) permissions for the KMM OperatorHub environment.
```

Author: changed

> ** *Consequence*: The KMM Operator Hub environment cannot build images.

Reviewer 2: ```suggestion
** *Consequence*: The KMM OperatorHub environment cannot build images.
```

Author: changed

** *Fix*: The RBAC permissions are updated to allow updating finalizers on the MIC resource, and then the appropriate rules created.

> ** *Result:* The KMM Operator Hub environment builds images without errors with the 'ManagedClusterModule' controller.

Reviewer 2: ```suggestion
** *Result:* The KMM OperatorHub environment builds images without errors with the `ManagedClusterModule` controller.
```

// MGMT-20861

> * The 'PreflightValidationOCP' custom resource, with a `kernelVersion: tesdt` causes the KMM Operator to panic.

Reviewer 2: ```suggestion
* The `PreflightValidationOCP` custom resource with a `kernelVersion: tesdt` causes the KMM Operator to panic.
```

Author: changed

> ** *Cause*: Creating a `PreflightValidationOCP` custom resource (CR), with a 'kernelVersion' flag that is set to 'tesdt', causes the Kernel Module Management (KMM) Operator to generate a panic runtime error.

Reviewer 2: ```suggestion
** *Cause*: Creating a `PreflightValidationOCP` custom resource (CR), with a `kernelVersion` flag that is set to `tesdt`, causes the Kernel Module Management (KMM) Operator to generate a panic runtime error.
```

** *Consequence*: Entering invalid kernel versions causes the KMM Operator to panic.

> ** *Fix*: A webhook - a method for one application to automatically send real-time data to another application when a specific event occurs - has been added to the `PreflightValidationOCP` CR.

Reviewer 2: ```suggestion
** *Fix*: A webhook - a method for one application to automatically send real-time data to another application when a specific event occurs - is now added to the `PreflightValidationOCP` CR.
```

Author: changed

> ** *Result:* The `PreflightValidationOCP` CR with invalid kernel versions can no longer be applied to the cluster, therefore, preventing the operator from generating a panic runtime error.

Reviewer 2: ```suggestion
** *Result:* The `PreflightValidationOCP` CR with invalid kernel versions can no longer be applied to the cluster, therefore, preventing the Operator from generating a panic runtime error.
```

// MGMT-20866
* The `PreFflightValidationOCP` custom resource, with a `kernelVersion` flag that is different that the one of the cluster, does not work.

> ** *Cause*: Creating a `PreflightValidationOCP` custom resource (CR), with a 'kernelVersion' flag that is different from the one of the cluster, does not work.

Reviewer 2: ```suggestion
** *Cause*: Creating a `PreflightValidationOCP` custom resource (CR), with a `kernelVersion` flag that is different from the one of the cluster, does not work.
```

Author: changed

** *Consequence*: The Kernel Module Management (KMM) Operator is unable to find the Driver Toolkit (DTK) input image for the new kernel version.

> ** *Fix*: Users should use the `PreflightValidationOCP` CR and explicitly set the 'dtkImage' field in the CR.

Reviewer 2: ```suggestion
** *Fix*: You must use the `PreflightValidationOCP` CR and explicitly set the `dtkImage` field in the CR.
```

** *Result:* Using the fields `kernelVersion` and `dtkImage` the feature can build installed modules for target {product-title} versions.

// MGMT-20888
* The KMM Operator version 2.4 documentation is updated with `PreflightValidationOCP` information.

** *Cause*: Previously, when creating an `PreflightValidationOCP` CR, you were required to supply the release-image. This has now changed and you need to set the `kernelVersion` the `dtkImage` fields.

** *Consequence*: The documentation was outdated and required an update.

** *Fix*: The documentation is updated with the new support details.

** *Result:* The KMM preflight feature is documented as expected.

> // MGMT-20892

Reviewer 2: Looks like a duplicate of this issue [here](https://github.com/openshift/openshift-docs/pull/95248/files#diff-a508e43343db975c5b8b7e3b4f30ef90441b9420f08df906454a6d85aa693514R89)

Author: You are correct. Deleted this instance. Thx! Good call!

> * Using imageRepoSecret in conjunction with DTK as imagestream gets authorization required error.

Reviewer 2: ```suggestion
* Using `imageRepoSecret` in conjunction with DTK as image stream gets authorization required error.
```

> ** *Cause*: On the Kernel Module Management (KMM) Operator, when you set the `imageRepoSecret` in the KMM module, and the build's resulting container image is defined to be stored in the cluster's internal registry, the build will fail to push the final image and generate an `authorization required` error.

Reviewer 2: ```suggestion
** *Cause*: On the Kernel Module Management (KMM) Operator, when you set the `imageRepoSecret` in the KMM module, and the build's resulting container image is defined to be stored in the cluster's internal registry, the build fails to push the final image and generate an `authorization required` error.
```

Author: changed

** *Consequence*: The KMM Operator does not work as expected.

** *Fix*: When the `imageRepoSecret` is user-defined, it is used as both a pull and push secret by the build process. In order to support using the cluster's internal regsitry, the auth token for that registry needs to be added to the `imageRepoSecret`. The token can be obtained from the "build" service account of the KMM module's namespace.

** *Result:* The KMM Operator works as expected.

> // test

Reviewer 2: ```suggestion

```

=== Known issues
// MGMT-20453 changed security settings in jira

> * The `ModuleUnloaded` event does not appear when a module is Unloaded.

Reviewer 2: Suggestion based on the assumption that unloaded is a state, if incorrect, please use the noun form
```suggestion
* The `ModuleUnloaded` event does not appear when a module is `Unloaded`.
```

> ** *Cause*: When a module is Loaded (using the create a 'ModuleLoad' event) or Unloaded (using the create a 'ModuleUnloaded' event) the events may not appear. This happens when you load and unload the kernel module in a quick succession.

Reviewer 2: ```suggestion
** *Cause*: When a module is `Loaded` (using the create a `ModuleLoad` event) or `Unloaded ` (using the create a `ModuleUnloaded` event) the events might not appear. This happens when you load and unload the kernel module in a quick succession.
```

Author: changed

> ** *Consequence*: The 'ModuleLoad' and 'ModuleUnloaded' events may not appear in {product-title}.

Reviewer 2: ```suggestion
** *Consequence*: The `ModuleLoad` and the `ModuleUnloaded` events might not appear in {product-title}.
```

Author: changed

> ** *Fix*: Alerting the user of this potential behavior and for awareness when working with modules.

Reviewer 2: ```suggestion
** *Fix*: Introduce an alerting mechanism for this potential behavior and for awareness when working with modules.
```

Author: changed

* In this release, KMM uses version 1.23 of the Golang programming language to ensure test continuity for partners.

// TELCODOCS-2197
* You can now schedule KMM pods by defining taints and tolerations. For more information, see xref:../hardware_enablement/kmm-kernel-module-management.adoc#kmm-using-tolerations-for-kernel-module-scheduling_kernel-module-management-operator[Using tolerations for kernel module scheduling].

[id="kmm-2-4-RN"]
== Release notes for Kernel Module Management Operator 2.4
=== New features
// TELCODOCS-2343

> * A new feature in this release is the addition of a preflight validation resource in the cluster that you can use to verify kernel modules to be installed on the nodes after cluster upgrades and possible kernel upgrades. Preflight validation also reports on the status and progress of each module in the cluster that it attempts or has attempted to validate. For more information, see xref:../updating/preparing_for_updates/kmm-preflight-validation.adoc#kmm-validation-kickoff_kmm-preflight-validation[Preflight validation for Kernel Module Management (KMM) Modules].

Author: fixed.

// TELCODOCS-2344

> * In this release, a new requirement when creating a kmod image is that both the `.ko` kernel module files and the `cp` binary must be included, which is required for copying files during the image loading process.

Reviewer 3: This requirement was added in KMM 2.3. We only extended the documentation to mention the requirement in this release.
